### ISSUE TITLE
How does inferior tty work?

### ISSUE THREAD

**Issue by hth313** [microsoft/vscode-cpptools#3951]
My debugger (gdb clone) handles debugee process terminal output using the inferior tty, but no such device is given to the debugger by VSCode (i.e. using the `-t` or `--tty` command line option).

I looked around, but I have not found how I can get VSCode to open and pass the inferior tty device name.

I use macOS 10.13.6, VSCode 1.36.1 and vscode-cpptools 0.24.0.
The debugger is my own, it implements the MI protocol.

**Comment by pieandcakes** [comment]
Hi @hth313 ,

If you want to see how it is implemented, it is within the [MIEngine](https://github.com/microsoft/miengine) project. I don't think we support exposing it to the user though.

The code you care about is [here](https://github.com/microsoft/MIEngine/blob/9372caff1e6b30a831da36290bc9d2821b010e99/src/MICore/UnixUtilities.cs#L40) where we take the launching `tty` and redirect it using shell parameters to `gdb`.

**Comment by hth313** [comment]
Hi @pieandcakes 

I looked at that code, but it does not seem to be the terminal I/O, but rather the shell used for gdb itself?

It is a bit confusing, as there can be up to 3 terminals (ttys) that are related to this. First of all the gdb process has its tty where the MI communication takes place. VS-Code also seems to have a bash terminal that is started in the root of the project directory. Finally, I think there should be a tty for the process being debugged. On Unix that is defined using `-t`, `--tty` command line options or using the `tty` command, and `inferior-tty` parameter. On Windows gdb has a `set new-console on/off` command.

Using find-grep on the VS-Code, I can find a place where it issues `set new-console on` in `MIDebugEngine/Engine.Impl/DebuggedProcess.cs:795`, that should open a new tty for the debugee, but that probably only works on Windows. I have yet to find where the tty is set for Unix, I cannot find much related to `tty` or `-t` at all.

References:
https://sourceware.org/gdb/current/onlinedocs/gdb/Input_002fOutput.html#index-inferior-tty
https://visualgdb.com/gdbreference/commands/tty
http://visualgdb.com/gdbreference/commands/set_new-console

**Comment by pieandcakes** [comment]
@hth313 [this](https://github.com/microsoft/MIEngine/blob/9372caff1e6b30a831da36290bc9d2821b010e99/src/MICore/UnixUtilities.cs#L63) line is where the debugger command is put together. We query for the running `tty` and store it in `$DbgTerm` variable and then pass it to the `-tty` flag when initializing `gdb`. 

This is done because we are working under the assumption that we execute the debugger command in the same terminal that is already being redirected either to the integrated console or the external console. 

To support your scenario, I think there would need to be a flag for the user to set and if that value is set, we would override the configuration and pass the user's `tty` to the parameter. Does that sound sufficient for your case? We would probably only offer it for users on Linux/macOs.

**Comment by hth313** [comment]
For some reason I do not understand, it seems my debugger does not get the `--tty` option as it says there is no `--tty` option given. (I run in lldb mode as gdb does not work for me, is that the reason?)

I think you misunderstand my scenario. I basically only want my debugger to work the way gdb does. I had so far implemented a separate tty, but I can see that gdb defaults to sharing std streams over the debugger console, so I just added support for that as well.

What I want is to understand how it works with VS Code, and if it wants to do the shared console then I am probably fine with that. The thing is that if you share the console with the debugger, I think there is a potential problem with stdin, as there seems to be no way to channel stdin over MI like there is with stdout target stream output.

So now I channel the target stdout output over MI and it looks like this in the Debug Console with engine logging enabled:

```
@"t"
1: (37921) ->@"h"
1: (37921) ->
@"h"
1: (37921) ->@"e"
1: (37921) ->
@"e"
1: (37921) ->@"r"
1: (37922) ->
@"r"
1: (37922) ->@"e"
@"e"
1: (37922) ->
1: (37922) ->@"!"
1: (37922) ->
@"!"
1: (37923) ->@"\n"
1: (37923) ->
@"\n"
```

If I run it without any logging, this is what ends up in the Debug Console:
```
@"H"
@"i"
@" "
@"t"
@"h"
@"e"
@"r"
@"e"
@"!"
@"\n"
```

I would have hoped it would print `Hi there!` followed by a newline.

The debugee program looks like this:

```
#include <stdio.h>

int main () {
  char *p = "Hi there!\n";
  while (*p) {
    putchar(*p++);
  }
  return 0;
}
```

**Comment by pieandcakes** [comment]
> (I run in lldb mode as gdb does not work for me, is that the reason?)

You can confirm what we send by enabling logging or while running, look at the process arguments we use to launch your debugger. When our logging first starts, we write the command line that we call at the top of the logging block.

> The thing is that if you share the console with the debugger, I think there is a potential problem with stdin, as there seems to be no way to channel stdin over MI like there is with stdout target stream output.

From what I know about MI, it doesn't seem to have a method to send commands directly to the debuggee. You would need to use the `tty` that is routed to the inferior. I don't know how `gdb` implements it since we are not allowed to view their source code due to legal restrictions.

[This](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Output-Syntax.html#GDB_002fMI-Output-Syntax) is the documentation that we follow and the code to parse it is [here](https://github.com/microsoft/MIEngine/blob/f6c7d337c10f9aec24788e908e3e43a4fa97a445/src/MICore/Debugger.cs#L1172).

I think your responses need to be prefaced by `~` and not `->` for output.

**Comment by hth313** [comment]
I had a look at how gdb behaves and also in the gdb source.

TL;DR; I see no `--tty` command line argument and I think you should  recognize `@` stream character.

I do not see any `--tty` when launching in the log:

```
--> E (output): {"event":"output","body":{"category":"console","output":"1: (140) LaunchOptions<LocalLaunchOptions xmlns='http://schemas.microsoft.com/vstudio/MDDDebuggerOptions/2014'\n"},"seq":2,"type":"event"}
1: (140) LaunchOptions<LocalLaunchOptions xmlns='http://schemas.microsoft.com/vstudio/MDDDebuggerOptions/2014'
--> E (output): {"event":"output","body":{"category":"console","output":"1: (151) LaunchOptions  ExePath='/Users/hth/projects/goabby/tmp/semihosting/puts'\n"},"seq":4,"type":"event"}
1: (151) LaunchOptions  ExePath='/Users/hth/projects/goabby/tmp/semihosting/puts'
--> E (output): {"event":"output","body":{"category":"console","output":"1: (151) LaunchOptions  WorkingDirectory='/Users/hth/projects/goabby/tmp/semihosting'\n"},"seq":6,"type":"event"}
1: (151) LaunchOptions  WorkingDirectory='/Users/hth/projects/goabby/tmp/semihosting'
--> E (output): {"event":"output","body":{"category":"console","output":"1: (151) LaunchOptions  ExeArguments=''\n"},"seq":8,"type":"event"}
1: (151) LaunchOptions  ExeArguments=''
--> E (output): {"event":"output","body":{"category":"console","output":"1: (152) LaunchOptions  MIMode='lldb'\n"},"seq":10,"type":"event"}
1: (152) LaunchOptions  MIMode='lldb'
--> E (output): {"event":"output","body":{"category":"console","output":"1: (152) LaunchOptions  MIDebuggerPath='/Users/hth/projects/goabby/bin/db6502'\n"},"seq":12,"type":"event"}
1: (152) LaunchOptions  MIDebuggerPath='/Users/hth/projects/goabby/bin/db6502'
--> E (output): {"event":"output","body":{"category":"console","output":"1: (152) LaunchOptions  WaitDynamicLibLoad='false'\n"},"seq":14,"type":"event"}
1: (152) LaunchOptions  WaitDynamicLibLoad='false'
--> E (output): {"event":"output","body":{"category":"console","output":"1: (153) LaunchOptions>\n"},"seq":16,"type":"event"}
1: (153) LaunchOptions>
--> E (output): {"event":"output","body":{"category":"console","output":"1: (153) LaunchOptions    <LaunchCompleteCommand>exec-run</LaunchCompleteCommand>\n"},"seq":18,"type":"event"}
1: (153) LaunchOptions    <LaunchCompleteCommand>exec-run</LaunchCompleteCommand>
--> E (output): {"event":"output","body":{"category":"console","output":"1: (153) LaunchOptions</LocalLaunchOptions>\n"},"seq":20,"type":"event"}
1: (153) LaunchOptions</LocalLaunchOptions>
--> E (output): {"event":"output","body":{"category":"console","output":"1: (286) Starting: \"/Users/hth/projects/goabby/bin/db6502\" --interpreter=mi\n"},"seq":22,"type":"event"}
1: (286) Starting: "/Users/hth/projects/goabby/bin/db6502" --interpreter=mi

```


When using the simple `putchar` program I showed above, the debugee output  under control of gdb comes intermixed with everything else, like there is no marker or anything, it just appears in the stream with everything else from the debugger:


```
c
&"c\n"
~"Continuing.\n"
^running
*running,thread-id="all"
(gdb) 
Hi there!
~"[Inferior 1 (process 30589) exited normally]\n"
```

If probably uses some kind of buffering here (line buffering or other) as it does not print immediately when a `putchar` appears.

This is probably acceptable when using the ordinary console interpreter when interacting with a human, but it is not ideal when using MI as the output potentially disturbs the output. However, it may work fine if you just copy whatever you do not understand to the console.. :)

The `@` target-stream-output appears to be for simulated system calls from a target (remote debugee). A remote debugee that is controlled by gdb can simulate a file system and std streams using gdb. I have no such setup, but that is what I understand from looking at the gdb source.

In my debugger I either have a remote system or a simulator (which is basically  like a simulated remote system), it is not a real Unix process. Thus, I use the `@` indicator for its stdout, but VS Code does not look for it as can be seen in the code you indicated, which is why the `@"X"` messages are shown instead of `X`.

The alternative way would be to provide a tty which somehow is connected to the debugee by the debugger. For a real Unix process, it is obvious how that works inside the debugger, it just provides the std file handles to the debugger process. For a remote target (or a simulated target like I have), the debugger needs to be informed by the remote target, either over the communications channel between the debugger and (target) debugee or like I do, by dealing with special breakpoints where the debugger figures out what the debugee wants, i.e. write to its stdout, which the debugger then handles on the behalf of the target debugee.

It think you should implement `@` and use that for the console stream, as it is part of the MI protocol and it seems like it is used in this situation for remote debugging (but I cannot prove it at this point as I lack a proper setup for that).

They way I hoped it would work is that you open a separate console for the debugee (at least on Unix) in VS Code and pass it to the debugger using `--tty`. That way it gets proper separation from the debugger console and can work for both input and output. I suspect that the corresponding `set new-console on/off` command on Windows  causes the debugger to open a console window for the debugee, so you basically get the same thing result, expect that it is probably not any way to integrate it among the tabs in VS Code. But, I am just an implementor of a debugger, not the actual user.
